[TypeChecker] Implement result builder transform via AST modification

xedin opened the pull request:
The transform changes closure body into a multi-statement closure
with all of the implicit result builder calls and type-checks it
like a regular closure. After transformation of a particular builder to a
particular function body is applied, it is cached and reused in case
other overloads require the same transform.

There are a couple of result builder specific changes mentioned below,
otherwise the logic to generate constraints and apply solutions is
unchanged:

- Placeholder variable: A variable declaration that doesn't have a
   type and infers it from first use site. If such a variable
   has an initializer, it would be type-checked during solution application.

 - `TypeJoinExpr` - an implicit expression that refers to a "join" variable
    and a set of expressions that should all produce the same type that
    becomes a type of a "join" variable.

Diagnostics: Since result builders are type-checked as regular multi-statement
closures it means that first error in the body skips the rest of it, which helps to
pinpoint issues and improve performance of an invalid code.

Resolves: rdar://85265414

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[BuilderTransform] Move some of the common logic into a base class

The base class is intended to facilitate logic sharing between
existing one-way constraint based transform and the new AST one.

xedin committed:
[Frontend] Add an option to enable new result builder transform

xedin committed:
[ConstraintSystem] Make it possible to storage `AnyFunctionRef` targets

xedin committed:
[AST] Add TypeJoin expression

This expression represents a variable and a set of expressions
that are type-checked together to form a type of the variable.

xedin committed:
[AST] NFC: Add a dedicated method to create implicit `Do` and `Brace`

xedin committed:
[CSSolver] Add support for `TypeJoinExpr`

xedin committed:
[CSClosure] Declarations prefixed with `$` can act as inferrable placeholders

If the type for such declaration could be completely inferred it
would be used during solution application. This is used to infer
types for result builder components.

xedin committed:
[ConstraintSystem] Switch `applySolutionToBody` to accept `AnyFunctionRef`

Result builder body could be either a closure or function e.g. accessor,
which means that solution application needs to handle both.

xedin committed:
[CSClosure] Generalize syntactic element simplification

Syntactic elements could be found in a closure or an accessor
which could also be a result builder.

xedin committed:
[ConstraintSystem] Allow placeholder vars to have initializer and nested holes

Initializers of such variables are going to be type-checked post factum,
when the type of the variable itself has been determined by the solver.

xedin committed:
[CSClosure] Teach syntactic element constraint generator about result builders

Closures and functions that have result builder transform applied have to be
handled specially by the constraint generator because transformed body is
associated with the context (function or closure) only during solution application.

xedin committed:
[CSClosure] NFC: Deprecate `isSupportedMultiStatementClosure`

A couple of places where it's still needed could use
`isInSingleExpressionClosure` instead, which is a safer
check because generator can now be used to handle
result builder transformed code and functions as well.

xedin committed:
[ConstraintSystem] Augment `AppliedBuilderTransform` to reference transformed body

If result builder AST transform is enabled, let builder transform descriptor
carry tranformed body for ease of access during solution application.

xedin committed:
[CSClosure] Conjunctions should reference result builder type variables

If builder type is not fully resolved, conjunction should reference
its variables otherwise it would get disconnected from required
contextual constraints.

xedin committed:
[CSClosure] `$__builderSelf` should be allowed to carry type variables

Just like parameter types, `$__builderSelf` is allowed to have not
yet inferred type variables.

xedin committed:
[CSDiagnostics] Don't try to diagnose empty case/switch statements as unsupported

Such statements are already incorrect and going to be diagnosed by the parser.

xedin committed:
[AST] Add a way to set parsed version of function/closure body via `AnyFunctionRef`

xedin committed:
[TypeChecker] Implement result builder transform via AST modification

The transform changes closure body into a multi-statement closure
with all of the implicit result builder calls and type-checks it
like a regular closure.

There are a couple of result builder specific changes mentioned below,
otherwise the logic to generate constraints and apply solutions is
unchanged:

- Placeholder variable: A variable declaration that doesn't have a
  type deduced and infers it from its first use. If such a variable
  has an initializer, it would be type-checked during solution application.

- TypeJoinExpr - an implicit expression that refers to a "join" variable
  and a set of expressions that should all produce the same type that
  becomes a type of a "join" variable.

xedin committed:
[BuilderTransform] Warn about absence of `buildLimitedAvailability(_:)`

xedin committed:
[ResultBuilder] Diagnostics: Don't diagnose unsupported statement if it is a brace

Brace statement are marked unsuported only if there is no appropriate overload
of `buildBlock` and that is diagnosed inline.

xedin committed:
[BuilderTransform] Cache previously applied transforms

Since result builder is just an AST transformation, the result
of a successful transform could be cache and reused with a different
`$__builderSelf` type.

xedin committed:
[TypeChecker] Enable result builder AST transform by default

While producing a combined solution, let's reflect the number of
fixes and holes discovered in the conjunction, that way it would
be possible to filter solutions and keep track of the fact that
there were issues in the conjunction.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
Reduced example from the issue found by the stress tester:

```swift
// RUN: %swift-ide-test -code-completion -source-filename %s -code-completion-token COMPLETE

protocol View {}

@resultBuilder
struct ViewBuilder {
  static func buildBlock(_ component: MyView) -> MyView {
    MyView()
  }
}

struct QStack<Content> {
  init(@ViewBuilder content: () -> Content) {}
  
  func qheet<Content>(@ViewBuilder content: @escaping () -> Content) -> some View {
    MyView()
  }
}

struct MyView : View {
  func qag<V>(_ tag: V) -> MyView {
    MyView()
  }
}


func foo() {
  QStack {
    MyView().qag(0)
  }
  .qheet() {
    MyView().#^COMPLETE^#
  }
}
```

used to produce a result `qag` but doesn’t produce any results with this PR.

xedin commented:
@ahoppen Thank you! I'll look into that

LucianoPAlmeida commented on lib/Sema/CSApply.cpp:
```diff
@@ -5121,6 +5121,11 @@ namespace {
     }
 
+    Expr *visitTypeJoinExpr(TypeJoinExpr *E) {
+      // Type-join expressions should not appear in the type-checked AST.
```
It is possible have a `TypeJoinExpr` here due to a bug? If it is maybe is worth assert or `llvm_unreachable` so it is noticeable?

xedin replied:
Yes, thanks for spotting this! Not having llvm_unreachable here is left-over from initial changes that introduced TypeJoinExpr

xedin replied:
Fixed!

LucianoPAlmeida commented on lib/Sema/CSStep.cpp:
```diff
@@ -958,24 +955,47 @@ StepResult ConjunctionStep::resume(bool prevFailed) {
+
+          // Transform all of the unbound outer variables into
+          // placeholders since we are not going to solve for
+          // each ambguous solution.
```
```suggestion
          // each ambiguous solution.
```

xedin replied:
Fixed!

xedin commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -723,178 +1475,167 @@ class BuilderClosureVisitor
+  UNSUPPORTED_STMT(PoundAssert)
+  UNSUPPORTED_STMT(Case)
 
 #undef CONTROL_FLOW_STMT
```
It should actually be only `#undef UNSUPPORTED_STMT` here

xedin replied:
Fixed, thank you!

xedin committed:
[CSStep] NFC: Fix a typo in a comment: ambguous -> ambiguous

xedin commented:
@ahoppen The reason why your example would no longer produce results is related to filtering - in code completion mode the filtering is disabled so first builder would end up with two solutions for `0` - `Int` and `Double` and that means that `MyView().qag(0)` marked as ambiguous and solver just skips the rest of the constraint system. I think the way to fix this would be to allow filtering everywhere in the builder body besides syntatic element that has code completion token, WDT?

ahoppen commented:
I see. I think your suggestion makes perfect sense and I already played around with it in https://github.com/apple/swift/pull/42024. I would be fine with merging the PR as-is and taking care of that problem later on.

xedin commented:
Sounds good!

xedin commented:
@swift-ci please build toolchain

ahoppen committed:
[CodeCompletion] Insert code completion token into build block call to retrieve contextual type

rdar://94224692

ahoppen committed:
[CodeCompletion] Drop ForCodeCompletion constraint system option for syntactic elements that don't contain completion token

xedin commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

xedin committed:
[TypeChecker] Stage syntactic statement diagnostics as warnings

These changes convert all errors produced by `performStmtDiagnostics`
into warnings when they are located in the body of a multi-statement
closure. This is required to stage enforcement that was missed by
original inference implementation which now also covers result buiders.

xedin commented:
@swift-ci please toolchain macosx

xedin commented:
@swift-ci please build toolchain macos

xedin commented:
@ahoppen Looks like changes to add source range to an implicit variable are causing verifier failures.

xedin committed:
[TypeChecker] NFC: Adjust improved result builder diagnostics

xedin committed:
[Diagnostics] Detect that return type mismatch belongs to a result builder

Since result builders are now type-checked like regular closures
diagnostics have to detect that contextual result type mismatch
occured in `return` statement associated with result builder transformed
function/closure.

ahoppen committed:
[TypeChecker] Use the brace statement as the start location of a buildLimitedAvailability call

Otherwise, the call to buildLimitedAvailability has a bigger source range than the brace statement that surrounds it, causing a verification failure.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

xedin committed:
[ResultBuilder] Don't run syntactic diagnostics on transformed do statements

Visiting of transformed `do` shouldn't run syntatic diagnostics because
they are going to be performed by `visit` that called `visitDoStmt`.

xedin commented:
@swift-ci please smoke test macOS

xedin commented:
@swift-ci Please SourceKit stress test

xedin commented:
Closing this one since we are going to land it piecemeal starting from https://github.com/apple/swift/pull/60065
